File: apps/pos_app/lib/data/models/human_authorization/assertion_v1.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:pos_app/data/models/human_authorization/canonical.dart';
import 'package:pos_app/data/models/human_authorization/error_codes.dart';
import 'package:pos_app/data/models/human_authorization/field_guards.dart';
import 'package:pos_app/data/models/human_authorization/staff_policy_epoch_v1.dart';

const assertionV1Schema = 'ohac.assertion.v1';

/// The only admissible trust level: software evidence from the enrolled
/// application sandbox. No hardware attestation is claimed in v1.
abstract final class OhacTrustLevel {
  static const applicationSandboxSoftware = 'APPLICATION_SANDBOX_SOFTWARE';
}

/// Immutable `ohac.assertion.v1` value object.
///
/// No `signature` field exists by design: the assertion is device-attributed
/// software evidence, never a claim of human non-repudiation. Because unknown
/// fields are rejected, adding one is a contract violation rather than a
/// silently ignored extra.
///
/// `permissionsUsed` is copied into an unmodifiable list at construction, so
/// callers cannot mutate validated data through the exposed reference and
/// later mutation of the original input list cannot alter this value object.
final class OhacAssertionV1 {
  final String schema;
  final String assertionId;
  final String tenantId;
  final String terminalId;
  final String deviceCredentialId;
  final String deviceCredentialVersion;
  final String epochSequence;
  final String epochDigest;
  final String authorizerUserId;
  final String operatorUserId;
  final String authorizerRole;
  final List<String> permissionsUsed;
  final String operationType;
  final String operationSchema;
  final String operationDigest;
  final String localAuthorizationSequence;
  final String localAuditId;
  final String localAuditEntryHash;
  final String posBuild;
  final String policySchema;
  final String trustLevel;
  final String authorizedAt;
  final String digest;

  OhacAssertionV1({
    required this.schema,
    required this.assertionId,
    required this.tenantId,
    required this.terminalId,
    required this.deviceCredentialId,
    required this.deviceCredentialVersion,
    required this.epochSequence,
    required this.epochDigest,
    required this.authorizerUserId,
    required this.operatorUserId,
    required this.authorizerRole,
    required List<String> permissionsUsed,
    required this.operationType,
    required this.operationSchema,
    required this.operationDigest,
    required this.localAuthorizationSequence,
    required this.localAuditId,
    required this.localAuditEntryHash,
    required this.posBuild,
    required this.policySchema,
    required this.trustLevel,
    required this.authorizedAt,
    required this.digest,
  }) : permissionsUsed = List<String>.unmodifiable(permissionsUsed);
}

const _assertionKeys = <String>[
  'schema',
  'assertionId',
  'tenantId',
  'terminalId',
  'deviceCredentialId',
  'deviceCredentialVersion',
  'epochSequence',
  'epochDigest',
  'authorizerUserId',
  'operatorUserId',
  'authorizerRole',
  'permissionsUsed',
  'operationType',
  'operationSchema',
  'operationDigest',
  'localAuthorizationSequence',
  'localAuditId',
  'localAuditEntryHash',
  'posBuild',
  'policySchema',
  'trustLevel',
  'authorizedAt',
  'digest',
];

final _utcRfc3339 = RegExp(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$');

/// Parses and validates an `ohac.assertion.v1` payload carrying an integrity
/// digest.
///
/// Order mirrors the TypeScript contract: canonicalize, digest equality,
/// exact keys, schema/trust identity, then field-level validation.
OhacResult<OhacAssertionV1> parseAssertionV1(Uint8List rawUtf8) {
  final canonical = canonicalizeOhac(rawUtf8);
  if (canonical case final OhacFailure<Uint8List> failure) {
    return OhacFailure<OhacAssertionV1>(failure.error);
  }

  Object? parsed;
  try {
    parsed = jsonDecode(utf8.decode(rawUtf8));
  } on FormatException {
    return OhacFailure(OhacError(OhacErrorCode.invalidJson));
  }
  final object = asObject(parsed);
  if (object == null) {
    return OhacFailure(OhacError(OhacErrorCode.invalidJson));
  }

  final transmittedDigest =
      isDigest(object['digest']) ? object['digest']! as String : null;
  if (transmittedDigest == null) {
    return OhacFailure(OhacError(OhacErrorCode.missingField, 'digest'));
  }
  final verified = verifyBodyDigest(object);
  if (ohacFailureAs<Map<String, dynamic>>(verified) case final verifiedFailure?) {
    return OhacFailure<OhacAssertionV1>(verifiedFailure.error);
  }

  final keys = requireExactKeys(object, _assertionKeys);
  if (ohacFailureAs<Map<String, dynamic>>(keys) case final keysFailure?) {
    return OhacFailure<OhacAssertionV1>(keysFailure.error);
  }

  if (object['schema'] != assertionV1Schema) {
    return OhacFailure(OhacError(OhacErrorCode.unsupportedSchema, 'schema'));
  }
  if (object['policySchema'] != staffPolicyEpochV1Schema) {
    return OhacFailure(
      OhacError(OhacErrorCode.unsupportedSchema, 'policySchema'),
    );
  }
  if (object['trustLevel'] != OhacTrustLevel.applicationSandboxSoftware) {
    return OhacFailure(OhacError(OhacErrorCode.invalidField, 'trustLevel'));
  }

  const uuidFields = <String>[
    'assertionId',
    'tenantId',
    'deviceCredentialId',
    'authorizerUserId',
    'operatorUserId',
    'localAuditId',
  ];
  for (final field in uuidFields) {
    if (!isLowercaseUuid(object[field])) {
      return OhacFailure(OhacError(OhacErrorCode.invalidField, field));
    }
  }
  const nonEmptyFields = <String>[
    'terminalId',
    'operationType',
    'operationSchema',
    'posBuild',
  ];
  for (final field in nonEmptyFields) {
    if (!isNonEmptyString(object[field])) {
      return OhacFailure(OhacError(OhacErrorCode.invalidField, field));
    }
  }
  const decimalFields = <String>[
    'deviceCredentialVersion',
    'epochSequence',
    'localAuthorizationSequence',
  ];
  for (final field in decimalFields) {
    if (!isDecimalString(object[field])) {
      return OhacFailure(OhacError(OhacErrorCode.invalidField, field));
    }
  }
  const digestFields = <String>[
    'epochDigest',
    'operationDigest',
    'localAuditEntryHash',
  ];
  for (final field in digestFields) {
    if (!isDigest(object[field])) {
      return OhacFailure(OhacError(OhacErrorCode.invalidField, field));
    }
  }
  if (!isOhacRole(object['authorizerRole'])) {
    return OhacFailure(OhacError(OhacErrorCode.invalidField, 'authorizerRole'));
  }
  final authorizedAt = object['authorizedAt'];
  if (authorizedAt is! String || !_utcRfc3339.hasMatch(authorizedAt)) {
    return OhacFailure(OhacError(OhacErrorCode.invalidField, 'authorizedAt'));
  }
  final permissionsUsed = object['permissionsUsed'];
  if (permissionsUsed is! List ||
      permissionsUsed.isEmpty ||
      permissionsUsed.any((permission) => !isNonEmptyString(permission))) {
    return OhacFailure(OhacError(OhacErrorCode.invalidField, 'permissionsUsed'));
  }
  final permissions = permissionsUsed.cast<String>();
  final sorted = requireSortedUnique(permissions, 'permissionsUsed');
  if (ohacFailureAs<List<String>>(sorted) case final sortedFailure?) {
    return OhacFailure<OhacAssertionV1>(sortedFailure.error);
  }

  return OhacSuccess(OhacAssertionV1(
    schema: assertionV1Schema,
    assertionId: object['assertionId']! as String,
    tenantId: object['tenantId']! as String,
    terminalId: object['terminalId']! as String,
    deviceCredentialId: object['deviceCredentialId']! as String,
    deviceCredentialVersion: object['deviceCredentialVersion']! as String,
    epochSequence: object['epochSequence']! as String,
    epochDigest: object['epochDigest']! as String,
    authorizerUserId: object['authorizerUserId']! as String,
    operatorUserId: object['operatorUserId']! as String,
    authorizerRole: object['authorizerRole']! as String,
    permissionsUsed: permissions,
    operationType: object['operationType']! as String,
    operationSchema: object['operationSchema']! as String,
    operationDigest: object['operationDigest']! as String,
    localAuthorizationSequence:
        object['localAuthorizationSequence']! as String,
    localAuditId: object['localAuditId']! as String,
    localAuditEntryHash: object['localAuditEntryHash']! as String,
    posBuild: object['posBuild']! as String,
    policySchema: object['policySchema']! as String,
    trustLevel: object['trustLevel']! as String,
    authorizedAt: authorizedAt,
    digest: transmittedDigest,
  ));
}

File: apps/admin_backend/src/modules/identity/human-authorization/contracts/assertion.v1.ts
import { canonicalizeOhac, verifyBodyDigest } from './canonical';
import {
  OHAC_ERROR_CODE,
  ohacFail,
  ohacOk,
  type OhacResult,
} from './error-codes';
import {
  asObject,
  isDecimalString,
  isDigest,
  isLowercaseUuid,
  isNonEmptyString,
  requireExactKeys,
  requireSortedUnique,
} from './field-guards';
import {
  OHAC_ROLE,
  STAFF_POLICY_EPOCH_V1_SCHEMA,
  type OhacRole,
} from './staff-policy-epoch.v1';

export const ASSERTION_V1_SCHEMA = 'ohac.assertion.v1' as const;

export const OHAC_TRUST_LEVEL = {
  APPLICATION_SANDBOX_SOFTWARE: 'APPLICATION_SANDBOX_SOFTWARE',
} as const;

export type OhacTrustLevel =
  (typeof OHAC_TRUST_LEVEL)[keyof typeof OHAC_TRUST_LEVEL];

export interface OhacAssertionV1 {
  readonly schema: typeof ASSERTION_V1_SCHEMA;
  readonly assertionId: string;
  readonly tenantId: string;
  readonly terminalId: string;
  readonly deviceCredentialId: string;
  readonly deviceCredentialVersion: string;
  readonly epochSequence: string;
  readonly epochDigest: string;
  readonly authorizerUserId: string;
  readonly operatorUserId: string;
  readonly authorizerRole: OhacRole;
  readonly permissionsUsed: readonly string[];
  readonly operationType: string;
  readonly operationSchema: string;
  readonly operationDigest: string;
  readonly localAuthorizationSequence: string;
  readonly localAuditId: string;
  readonly localAuditEntryHash: string;
  readonly posBuild: string;
  readonly policySchema: string;
  readonly trustLevel: OhacTrustLevel;
  readonly authorizedAt: string;
  readonly digest: string;
}

/**
 * No `signature` field exists by design: the assertion is device-attributed
 * software evidence, never a claim of human non-repudiation. Because unknown
 * fields are rejected, adding one is a contract violation rather than a
 * silently ignored extra.
 */
const ASSERTION_KEYS = [
  'schema',
  'assertionId',
  'tenantId',
  'terminalId',
  'deviceCredentialId',
  'deviceCredentialVersion',
  'epochSequence',
  'epochDigest',
  'authorizerUserId',
  'operatorUserId',
  'authorizerRole',
  'permissionsUsed',
  'operationType',
  'operationSchema',
  'operationDigest',
  'localAuthorizationSequence',
  'localAuditId',
  'localAuditEntryHash',
  'posBuild',
  'policySchema',
  'trustLevel',
  'authorizedAt',
  'digest',
] as const;

const UTC_RFC3339 = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;

const isRole = (value: unknown): value is OhacRole =>
  typeof value === 'string' &&
  (Object.values(OHAC_ROLE) as readonly string[]).includes(value);

const isTrustLevel = (value: unknown): value is OhacTrustLevel =>
  value === OHAC_TRUST_LEVEL.APPLICATION_SANDBOX_SOFTWARE;

export const parseAssertionV1 = (
  rawUtf8: Buffer,
): OhacResult<OhacAssertionV1> => {
  const canonical = canonicalizeOhac(rawUtf8);
  if (canonical.ok === false) return canonical;

  let parsed: unknown;
  try {
    parsed = JSON.parse(rawUtf8.toString('utf8'));
  } catch {
    return ohacFail(OHAC_ERROR_CODE.INVALID_JSON);
  }
  const object = asObject(parsed);
  if (object === undefined) {
    return ohacFail(OHAC_ERROR_CODE.INVALID_JSON);
  }

  const transmittedDigest = isDigest(object.digest) ? object.digest : undefined;
  if (transmittedDigest === undefined) {
    return ohacFail(OHAC_ERROR_CODE.MISSING_FIELD, 'digest');
  }
  const verified = verifyBodyDigest(object);
  if (verified.ok === false) return verified;

  const keys = requireExactKeys(object, ASSERTION_KEYS);
  if (keys.ok === false) return keys;

  if (object.schema !== ASSERTION_V1_SCHEMA) {
    return ohacFail(OHAC_ERROR_CODE.UNSUPPORTED_SCHEMA, 'schema');
  }
  if (object.policySchema !== STAFF_POLICY_EPOCH_V1_SCHEMA) {
    return ohacFail(OHAC_ERROR_CODE.UNSUPPORTED_SCHEMA, 'policySchema');
  }
  if (!isTrustLevel(object.trustLevel)) {
    return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, 'trustLevel');
  }

  for (const field of [
    'assertionId',
    'tenantId',
    'deviceCredentialId',
    'authorizerUserId',
    'operatorUserId',
    'localAuditId',
  ] as const) {
    if (!isLowercaseUuid(object[field])) {
      return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, field);
    }
  }
  for (const field of [
    'terminalId',
    'operationType',
    'operationSchema',
    'posBuild',
  ] as const) {
    if (!isNonEmptyString(object[field])) {
      return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, field);
    }
  }
  for (const field of [
    'deviceCredentialVersion',
    'epochSequence',
    'localAuthorizationSequence',
  ] as const) {
    if (!isDecimalString(object[field])) {
      return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, field);
    }
  }
  for (const field of [
    'epochDigest',
    'operationDigest',
    'localAuditEntryHash',
  ] as const) {
    if (!isDigest(object[field])) {
      return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, field);
    }
  }
  if (!isRole(object.authorizerRole)) {
    return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, 'authorizerRole');
  }
  if (
    typeof object.authorizedAt !== 'string' ||
    !UTC_RFC3339.test(object.authorizedAt)
  ) {
    return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, 'authorizedAt');
  }
  if (
    !Array.isArray(object.permissionsUsed) ||
    object.permissionsUsed.length === 0 ||
    object.permissionsUsed.some((permission) => !isNonEmptyString(permission))
  ) {
    return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, 'permissionsUsed');
  }
  const permissionsUsed = object.permissionsUsed as readonly string[];
  const sorted = requireSortedUnique(permissionsUsed, 'permissionsUsed');
  if (sorted.ok === false) return sorted;

  return ohacOk({
    schema: ASSERTION_V1_SCHEMA,
    assertionId: object.assertionId as string,
    tenantId: object.tenantId as string,
    terminalId: object.terminalId as string,
    deviceCredentialId: object.deviceCredentialId as string,
    deviceCredentialVersion: object.deviceCredentialVersion as string,
    epochSequence: object.epochSequence as string,
    epochDigest: object.epochDigest as string,
    authorizerUserId: object.authorizerUserId as string,
    operatorUserId: object.operatorUserId as string,
    authorizerRole: object.authorizerRole,
    permissionsUsed,
    operationType: object.operationType as string,
    operationSchema: object.operationSchema as string,
    operationDigest: object.operationDigest as string,
    localAuthorizationSequence: object.localAuthorizationSequence as string,
    localAuditId: object.localAuditId as string,
    localAuditEntryHash: object.localAuditEntryHash as string,
    posBuild: object.posBuild as string,
    policySchema: object.policySchema,
    trustLevel: object.trustLevel,
    authorizedAt: object.authorizedAt,
    digest: transmittedDigest,
  });
};
